### PR TITLE
Implement SnowflakeObject SlashArgumentConverter

### PR DESCRIPTION
**Pull request by ecrocombe**
# Summary
Implement the use of `SnowflakeObject` in Slash command arguments for the `DSP.Commands` lib

# Details
Created mostly out of necessity for an individual project, this takes a note from `DSP.SlashCommands` `SlashCommandsExtension`, and implements the converter in a similar fashion.

# Changes proposed
* With limitations on the number of SlashCommands able to be registered, making a new command to take either a Role or DiscordMember seems unnecessary and a waste of our limited number of commands. This will allow users to be able to cast the SnowflakeObject into either a DiscordUser, DiscordMember or DiscordRole and be able to either perform an action on a single user, or all users in a role, much like how it was historically done in DSP.SlashCommands.

# Notes
No testing has been performed, at all, what-so-ever. 
This is more of a "please make this work, I need it" kind of PR.

### CODE REVIEW

File: DSharpPlus.Commands/Converters/DiscordSnowflakeObjectConverter.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        }
+        else if (eventArgs.Interaction.Data.Resolved.Users != null && eventArgs.Interaction.Data.Resolved.Users.TryGetValue((ulong)option.Value, out DiscordUser? user))
+        {
+            return Task.FromResult(Optional.FromValue<SnowflakeObject>(user));
+        }
+        else
+        {
+            this._logger.LogError("Failed to resolve SnowflakeObject type."); 
+            return Task.FromResult(Optional.FromNoValue<SnowflakeObject>());
+        }
```
Comment (OoLunar) [inline DSharpPlus.Commands/Converters/DiscordSnowflakeObjectConverter.cs:42]: @akiraveliara Should we grab the existing converters and use those instead?

**Review by OoLunar — CHANGES_REQUESTED** [review]
See first review comment, please stand by for the second review command to be peer reviewed.

File: DSharpPlus.Commands/Converters/DiscordSnowflakeObjectConverter.cs (hunk 1 above)
Comment (akiraveliara) [reply DSharpPlus.Commands/Converters/DiscordSnowflakeObjectConverter.cs:42]: probably a good idea, if achievable without causing disaster

**Comment by OoLunar** [comment]
I apologize for spamming your notifications

**Review by OoLunar — APPROVED** [review]

**Comment by ecrocombe** [comment]
InteractionConverterContext.Interaction.Data.Resolved.Roles can be null when mentioning a DiscordMember or DiscordRole, causing NRE.

Suggest either a null check on the individual used ISlashArgumentConverter<>'s or implement custom logic for ISlashArgumentConverter<SnowflakeObject>

**Pull request merged**